### DISCUSSION TITLE
FLTK 1.4 Windows, Screen scaling, rounding issues with Fl_Box drawing

### DISCUSSION THREAD

**Discussion by dannye** [fltk/fltk#1089] [General]
**Problem:**

In all my previous testing with FLTK's screen scaling, I only considered using the Ctrl- and Ctrl+ keyboard shortcuts. I assumed those shortcuts were the only way to set FLTK's internal scaling factor to anything other than 100%.

It didn't even cross my mind until now to use this scaling option on Windows:
![image](https://github.com/user-attachments/assets/6cb099c2-4408-48bb-9c38-068df2623081)

I assumed that this^ scaling would still be performed by the OS and not by FLTK. I realize now that this assumption is not correct.

It was my intention to disable FLTK's internal scaling with `Fl::keyboard_screen_scaling(0);` but now I see that is not sufficient.

Here are the problems that causes.
(Note: For consistency, all screenshots of FLTK 1.3 will use the _white_ theme and all screenshots of FLTK 1.4 will use the _gray_ theme.)

In a previous release of my editor that used FLTK 1.3, here is what the window looks like when the Windows scale setting is set to 125%:
![image](https://github.com/user-attachments/assets/9cc5e69b-a798-4804-b333-998bfc9d6d54)

Most edges and text are a bit blurry, but everything is exactly where you would expect.

But here is the most recent version of my editor that uses FLTK 1.4 when the Windows scale setting is set to 125%:
![image](https://github.com/user-attachments/assets/6b179ea1-27c6-4b6c-85f8-108f0e0e4488)

The x, y, w, and h of all Fl_Boxes have a rounding error of 0 to 1.

**Close-ups:**

This means that "filled" boxes are often not actually filled:
![image](https://github.com/user-attachments/assets/65f3fb71-7b98-43ce-93b5-81013b8da97f)

![image](https://github.com/user-attachments/assets/160de808-78c5-4d6e-b042-f92c16cc291c)

And there are gaps between adjacent boxes, allowing you to see behind the boxes, which you're not supposed to be allowed to see:
![image](https://github.com/user-attachments/assets/bb7b2ee1-76eb-4faa-aa9e-23d1483dc71b)

And there are gaps between "concentric" rectangles:
![image](https://github.com/user-attachments/assets/30825851-9a54-4233-80aa-4aa35129ad61)

Additionally, xpm images are scaled very poorly.

This is what an icon normally looks like at 100%:
![image](https://github.com/user-attachments/assets/2f4a7ea8-3de2-4586-8ba0-1032149a6e13)

This is what the icon looks like in FLTK 1.3 at 125%: (ignore the fact that the artwork itself is slightly outdated here)
![image](https://github.com/user-attachments/assets/8e54fc22-e126-417f-ab22-eed9c3e813fc)

And this is what the icon looks like in FLTK 1.4 at 125%:
![image](https://github.com/user-attachments/assets/33a8b72e-f1da-4efe-beff-a5ef14fee946)

Note: For full transparency, I have `FLTK_GRAPHICS_GDIPLUS` disabled for my program on Windows, which I'm pretty sure is partially related to the xpm scaling***. But I have it disabled because GDI Plus drawing is just too slow, particularly with regard to Fl_Menubars.

*** EDIT: I think I was wrong about this. I tested with `FLTK_GRAPHICS_GDIPLUS` enabled again, and the xpm scaling was still just as pixelated. I must be misremembering that GDI Plus automatically smoothed the upscaling of xpm icons.

**Questions:**

In addition to `Fl::keyboard_screen_scaling(0);` is there something else I can do to disable **all** FLTK internal scaling?

If not, can these rounding errors be fixed? Can the xpm scaling be fixed? (I understand if these turn into cans-of-worms.)

**More Thoughts:**

Now that I realized that `Fl::keyboard_screen_scaling(0);` is not sufficient to avoid FLTK scaling, that means the outstanding issues from #919 still affect my program, even though I thought I completely avoided them.

Most significantly, the issue that a window can be dragged (or simply moved with Win+Shift+Left/Right) from one monitor with one scaling factor and dropped against the top edge of another monitor with a different scaling factor and then the drawing of that window will be ruined.

This issue alone is enough for me to strongly prefer the first option (disable **all** FLTK internal scaling). I much rather the window be a tiny bit blurry but always drawn correctly.

I bring all this up because a user of my program sent me a screenshot of the editor showing all the badly rounded rectangles, which was a complete shock to me, and it turns out they use the above-mentioned Windows scaling setting to set their two monitors to two different scales.

Thanks for reading and sorry for the wall of text.

**Comment by ManoloFLTK** [comment]
I believe the issue you report is visible using FLTK's test program `boxtype`. The `FL_BORDER_BOX` has some unfilled pixels in its inside. I will try to fix this, but, before that, I'd like to make these points:
- FLTK 1.4 has been extensively tested with option `FLTK_GRAPHICS_GDIPLUS` turned on, and in that case, the issue discussed here with program `boxtype` does not occur, and presumably not with your program either.
- I'm very surprised by your argument to turn off `FLTK_GRAPHICS_GDIPLUS` because it slows down menu drawing. FLTK's menubar test application uses quite elaborate menu objects, including an extremely long menu, and I perceive it quite responsive with `FLTK_GRAPHICS_GDIPLUS` turned on.
- no issue should occur with scale factor values a multiple of 100%.

In general terms, setting FLTK's scaling factor to 125% has exactly the same effect as setting the system scale to 125% using Windows `Settings` application. The whole purpose of this is to get rid in FLTK 1.4 of the blurring you report when using FLTK 1.3 and system scaling values > 100%.

About scaling xpm icons when they get drawn. FLTK 1.4  does this:
- consider an icon of` W x H `pixels and a scale factor of 1.25 (corresponding to 125% in Settings);
- use `Fl_Pixmap::copy()` to produce an xpm image of size `int(1.25*W) x int(1.25*H)` pixels;
- draw this scaled icon as if the current scaling factor was at 1.00.

The goal of this is that pixmaps are not blurred when they are scaled up or down, and that's what you observe. But scaling up a pixmap to a non integer scale factor of `s` is challenging: how to color `N*s*s` pixels starting from information about how to color `N` pixels? I'm receptive to any suggestion about how to do that better. A working solution today is to use larger pixmaps (say, twice larger) and to give them a drawing size equal to that of the initial pixmap. FLTK 1.4 will then scale them down rather than up to draw them for scale values between 100% and 200%. Another approach would be to have FLTK at run-time convert small pixmaps to `Fl_RGB_Image` objects (a constructor of class `Fl_RGB_Image` does exactly this) and use these as widget images. They will therefore be scaled up with GDIplus image scaling algorithm.

**Reply by MatthiasWM** [reply]
@ManoloFLTK already described very well what the issues are. It's hard to impossible in the integer world of pixels to scale by a factor that is not a multiple of 1 (100%). The solution for pixmaps is quite satisfying though: Store all your icons at 200% resolution and let FLTK scale them down to 125% instead of up from 100%. It works so much better!

**Reply by dannye** [reply]
Thanks for all the tips. I will post some examples of what I mean by the slow menubar drawing with GDI Plus in a separate reply thread below in this discussion in just a moment.

**Reply by dannye** [reply]
> * FLTK 1.4 has been extensively tested with option `FLTK_GRAPHICS_GDIPLUS` turned on, and in that case, the issue discussed here with program `boxtype` does not occur, and presumably not with your program either.

Hm, this does not appear to be true in my testing.

Here is what it looks like with `FLTK_GRAPHICS_GDIPLUS` enabled and with the Windows Settings app scale set to 125%:

![image](https://github.com/user-attachments/assets/c262309a-a252-4367-bbb5-7ae8167ad350)

(The giveaway that GDI Plus is definitely enabled is the "smooth" arrows in the scrollbar buttons that are visible in that screenshot. Without GDI Plus, those arrows are a little "pixelated". So GDI Plus is definitely enabled, but FL_Boxes are still not reliably filled.)

(I also must be misremembering that GDI Plus would automatically smooth the icons when upscaling the xpm images. The icons are still just as blocky. So enabling GDI Plus neither hurts nor helps any of the issues described in the OP.)

(I'm also fairly confident that this "unfilled rect" issue was not an issue when I originally did the Ctrl- / Ctrl+ scaling testing at the beginning of this year. Surely I would have noticed it if it was a problem.)

**Reply by dannye** [reply]
> give them a drawing size equal to that of the initial pixmap.

> Store all your icons at 200% resolution and let FLTK scale them down to 125% instead of up from 100%.

Sorry for being inept, but how do I actually do this?

I tried doubling the dimensions of the xpm image, but now it just draws like this, centered and cropped inside its Fl_Button:

![image](https://github.com/user-attachments/assets/9621b5c2-49c5-4db9-a8fa-e5d02533217f)

I can set the bounding box to a centered 16x16 area by using the right overload of `Fl_Widget::draw_label(int, int, int, int)` instead of `Fl_Widget::draw_label()` but now it's just cropped more:

![image](https://github.com/user-attachments/assets/b11404f3-5514-49f5-a856-e074104fde73)

I'm not sure how to let FLTK scale it down. I tried to look at things related to Widget labels, like Fl_Labeltype and Fl_Align to find a "scale to fit option" but nothing sticks out to me.

**Reply by ManoloFLTK** [reply]
You have to use Fl_Image::scale(), new in FLTK 1.4

**Reply by MatthiasWM** [reply]
Oh, sorry, yes. Say the original image was 24x24. So you scale and refine your image to 48x48 and call
  MyButton->image(myIcon48x48);
  MyButten->image()->scale(24, 24, 0, 1);

0 is for non-proportional scaling if needed, and 1 enables scaling up, too, so a 300% scale will be supported too.

**Comment by Albrecht-S** [comment]
As an additional note, from faint memory: isn't there an option to create and apply a (Windows) "manifest" that makes an app "not DPI-aware" (or whatever it's called) so Windows does the entire scaling? ISTR that FLTK does not scale in this case, and that's what @dannye wanted to achieve in the first place. Unfortunately I don't find any info about this except:

[Our docs](https://www.fltk.org/doc-1.4/osissues.html#osissues_windows_highdpi) say: _"FLTK apps may also use the manifest mechanism to declare their level of DPI awareness. The FLTK library adapts to the DPI awareness level set in the app's manifest, which can be lower than the default level if the manifest sets it so."_

**Comment by dannye** [comment]
So with GDI Plus enabled on Windows, "scrubbing" the mouse cursor up and down a menu causes lots of flickering that is not present when GDI Plus is disabled.

These 3 examples are of _sequential_ frames (screen grabbed from a video editor).

Example 1, moving the mouse up and down over some menu items:

<details>

![Screenshot 2024-10-13 112148](https://github.com/user-attachments/assets/09c662f0-731b-4cd3-85b3-4067e652f8c2)
![Screenshot 2024-10-13 112203](https://github.com/user-attachments/assets/3863bfc7-88f9-44c0-aa87-1bdac59e5c5e)
![Screenshot 2024-10-13 112214](https://github.com/user-attachments/assets/a5707596-72c8-4972-bbe9-86c7533a1cf9)
![Screenshot 2024-10-13 112238](https://github.com/user-attachments/assets/0a320886-6f6f-423e-a910-8b5f3516dbe7)
![Screenshot 2024-10-13 112257](https://github.com/user-attachments/assets/078eeb82-8ea6-4307-8c4a-d0cdf09d63bf)
![Screenshot 2024-10-13 112312](https://github.com/user-attachments/assets/f58d5036-3f15-432a-b212-84fcf896a42d)
![Screenshot 2024-10-13 112325](https://github.com/user-attachments/assets/665eb745-4b9d-434c-bb84-9d1592616eb9)

</details>

Example 2,  opening a submenu by mousing over it:

<details>

![Screenshot 2024-10-13 113537](https://github.com/user-attachments/assets/a698ce43-d897-4250-8023-019f938e27bc)
![Screenshot 2024-10-13 113551](https://github.com/user-attachments/assets/9f9598bd-b7bc-4cc7-a397-61dd5f9008f0)
![Screenshot 2024-10-13 113601](https://github.com/user-attachments/assets/eedf5faf-b52e-4a4c-82f7-7324b2e4f27a)
![Screenshot 2024-10-13 113617](https://github.com/user-attachments/assets/8339a66c-c6d5-48a5-9b05-cd797a34c966)
![Screenshot 2024-10-13 113629](https://github.com/user-attachments/assets/9258ac02-77cf-4099-bdf9-0a540108d05d)

</details>

Example 3, moving the mouse left and right over the top menubar:

<details>

![Screenshot 2024-10-13 114152](https://github.com/user-attachments/assets/441f78e0-1901-4917-be2c-1fa3c13949c4)
![Screenshot 2024-10-13 114207](https://github.com/user-attachments/assets/82562db1-61ce-4bf6-989f-61a254d10797)
![Screenshot 2024-10-13 114223](https://github.com/user-attachments/assets/c65918c4-b895-4796-83f2-f1d8b34dfb0a)
![Screenshot 2024-10-13 114231](https://github.com/user-attachments/assets/808f00f2-2591-4111-8afa-248cacf3a9fa)
![Screenshot 2024-10-13 114242](https://github.com/user-attachments/assets/ad7d3e4b-91f7-4fcb-b579-d14d620fc60f)
![Screenshot 2024-10-13 114257](https://github.com/user-attachments/assets/a672d585-02fd-4b63-a89a-03d5c8b31a01)
![Screenshot 2024-10-13 114308](https://github.com/user-attachments/assets/0428c52c-e8d9-42ba-958f-148fd8b4dd1f)
![Screenshot 2024-10-13 114323](https://github.com/user-attachments/assets/9c5a500f-3d88-458b-8bdc-76b72c4a88bf)

</details>


A few more independent frames (not sequential) showing text partially or completely missing:

<details>

![Screenshot 2024-10-13 113330](https://github.com/user-attachments/assets/40d66f79-0b75-490c-b33f-be9e7c8d9ac7)
![Screenshot 2024-10-13 113411](https://github.com/user-attachments/assets/763ba9d8-3798-415c-a47f-3d34c14e63c0)
![Screenshot 2024-10-13 113423](https://github.com/user-attachments/assets/3e73d3bb-6263-4172-a4f9-5558b63c85ec)
![Screenshot 2024-10-13 113452](https://github.com/user-attachments/assets/c499de46-4462-4a27-8cea-91b4d1c94b36)
![Screenshot 2024-10-13 113723](https://github.com/user-attachments/assets/b319e951-cefc-4991-aa90-c5cf6a078ec3)
![Screenshot 2024-10-13 115015](https://github.com/user-attachments/assets/a18f1717-01ee-4bdc-830b-9ecb3f3147ee)
![Screenshot 2024-10-13 115030](https://github.com/user-attachments/assets/1e9b1d70-6318-487b-9889-a322ba7d93b3)
![Screenshot 2024-10-13 115045](https://github.com/user-attachments/assets/8a20eb97-2ad2-422f-a02f-b1c14d05c742)

</details>

The flickering is definitely noticeable to the naked eye and quite visually distracting for me. The issue happens in both debug and release builds (maybe just _slightly_ less frequently in release builds, but really not by much). I cannot reproduce _any_ of these artifacts with GDI Plus disabled.

**Reply by ManoloFLTK** [reply]
I can't reproduce here under Windows 11 arm64 what you report.

**Comment by ManoloFLTK** [comment]
Please, try applying with `git apply` the attached patch. It should fix all boxes of type FL_BORDER_BOX that I believe are your filled rectangles. The gaps between adjacent rectangles should also disappear. The fix should be visible both with and without GDIplus. What you called concentric rectangles will remain with an issue. Report here what you get and give an idea of what FLTK statements draw remaining elements with issues, if any.

[rectbound.patch.txt](https://github.com/user-attachments/files/17357018/rectbound.patch.txt)

**Reply by dannye** [reply]
Thanks @ManoloFLTK.

From what I can tell, the "unfilled" rectangle issue is completely fixed. ✅ 

Regarding the "gap between adjacent rectangles" issue, the gaps are definitely gone but I think it was a slight over-correction.
At a regular 100% scale, the `w` and `h` of each box is 1 greater than expected. (The x and y of all boxes is always correct.)

It now looks like this (scaled up x4):

![image](https://github.com/user-attachments/assets/e091940a-8108-461f-8101-28db590da8a2)

It ordinarily should look like this:

![image](https://github.com/user-attachments/assets/9409b33c-59fe-4b62-b8a3-bd829a81e3b6)

And I completely admit the "gap between concentric rectangles" issue is more of a "me problem."
(When I was originally learning FLTK 2 years ago, I "lazily" made a few Fl_Box subclasses with a custom `::draw()` that would just draw 1 or 2 extra "inner" rectangles for all the cases where I want a 2-or-3-pixel-thick rectangle for both `FL_BORDER_BOX` and `FL_BORDER_FRAME` ([example](https://github.com/dannye/crystal-tracker/blob/bcd6e99eeaac94ce7b20d536d15e801059991ca4/src/piano-roll.cpp#L56-L60)). I should probably just come up with something better for that though.)

**Reply by ManoloFLTK** [reply]
About concentric rectangles: you could try this to draw a 2 pixel-thick rectangle:
```
fl_line_style(FL_SOLID,2); 
fl_rect(x,y,w,h);
fl_line_style(FL_SOLID);
```

**Reply by ManoloFLTK** [reply]
About the slight over-correction. Yes, there was one. This will remove it:
Consider line 779 of patched file `src/Fl_Graphics_Driver.cxx` which reads
`                     this->floor(x + w) - this->floor(x), this->floor(y + h) - this->floor(y), `
Modify it as follows
`                     this->floor(x + w `**- 1**`) - this->floor(x), this->floor(y + h` **- 1**`) - this->floor(y),`

**Reply by dannye** [reply]
Unfortunately that change brings back the gaps when at 125%:
![image](https://github.com/user-attachments/assets/838aafa0-1b8f-41f3-9d3c-be00dfd1e9cb)

**Reply by dannye** [reply]
> About concentric rectangles: you could try this to draw a 2 pixel-thick rectangle:
> 
> ```
> fl_line_style(FL_SOLID,2); 
> fl_rect(x,y,w,h);
> fl_line_style(FL_SOLID);
> ```

I may tinker with this, but the reason I went with the "lazy" approach is that I need a guarantee that as the border gets thicker, the border only grows inward (ie, a 2-pixel-wide or 3-pixel-wide border never makes the box any taller or wider). And I'd be concerned that simply increasing the line style thickness wouldn't provide me with that guarantee (although I haven't experimented with it much).

There are also cases where I don't use the same color for all 3 "rings" of the border:
![image](https://github.com/user-attachments/assets/ba96137e-27e3-4f8a-a605-380004a3ba38)

Unsurprisingly, it doesn't look great at 125% scale:
![image](https://github.com/user-attachments/assets/16026850-675d-48bc-85bb-acc6a93a8bce)

I could experiment with replacing the 2 white rects with a single 2-pixel-thick rectangle, but I'm unsure how to guarantee that the white rect always "hugs" the black outer rect.

**Reply by dannye** [reply]
I also call my approach "lazy" because I assume that a non-lazy approach would be to make a custom Fl_Boxtype + draw function, I just haven't looked into that yet.

**Reply by dannye** [reply]
> Unsurprisingly, it doesn't look great at 125% scale:
> ![image](https://github.com/user-attachments/assets/16026850-675d-48bc-85bb-acc6a93a8bce)

I'd like to point out one thing about this screenshot; You can see that the x, y, and w of that box is exactly correct, but the h should have been 1 greater. You can see there is a medium-gray "background row" behind the box, and that gray row extends 1 pixel below the box. That gray rectangle is drawn directly with `fl_rectf(x, y, w, h, c)` and it is always scaled exactly correctly (there are never any gaps between the "row" rectangles drawn with `fl_rectf`) so I think all the logic for precisely correct rectangle rounding already exists in FLTK, but something about `Fl_Box::draw()` / `draw_box()` is missing something that `fl_rectf` has.

(sorry for quadruple posting)

**Reply by ManoloFLTK** [reply]
> Unfortunately that change brings back the gaps when at 125%:

OK. Yes, I can see that. Please, try this other modification of the patched file that works well in my tests:

Consider line 779 of patched file src/Fl_Graphics_Driver.cxx which reads
`  this->floor(x + w) - this->floor(x), this->floor(y + h) - this->floor(y),`
Modify it as follows
`  this->floor(x + w) - int(scale()) - this->floor(x), this->floor(y + h) - int(scale()) - this->floor(y),`

**Reply by dannye** [reply]
@ManoloFLTK That works great, thank you. All adjacent rectangles are perfectly flush.

**Reply by dannye** [reply]
One other lazy/hacky thing that I do is for the high contrast theme, after drawing certain rectangles, I immediately redraw the frame in a different color (wasteful overdraw, because I couldn't figure out an easier way to draw a Fl_Box with a custom frame color _and_ a custom fill color).

Like this:

```cpp
void Key_Box::draw() {
	draw_box();
	if (OS::current_theme() == OS::Theme::HIGH_CONTRAST)
		draw_box(FL_BORDER_FRAME, fl_darker(FL_SELECTION_COLOR));
	draw_label();
}
```

I would expect the "second frame" to overlap perfectly with the first frame, but that is not always the case. `w` and `h` are sometimes short by 1 on the second draw:

![image](https://github.com/user-attachments/assets/42dedcaf-d53e-476b-88e2-ed4908ecfa1b)

Of course, I know your patch is not final, so this is just an FYI.

(It looks like both `draw_box()` and `draw_box(Fl_Boxtype, Fl_Color)` call `draw_box(t, x, y, w, h, c)` with the same x/y/w/h, so I would guess the discrepancy comes from FL_BORDER_BOX versus FL_BORDER_FRAME.)

**Reply by dannye** [reply]
Yeah, applying your same fix to `Fl_Scalable_Graphics_Driver::rect()` like you did for `rectbound()` appears to fix it:

```diff
 void Fl_Scalable_Graphics_Driver::rect(int x, int y, int w, int h)
 {
   if (w > 0 && h > 0) {
-    int s = (int)scale()/2;
-    rect_unscaled(this->floor(x) + s, this->floor(y) + s,
-                  this->floor(x + w - 1) - this->floor(x),
-                  this->floor(y + h - 1) - this->floor(y));
+    rect_unscaled(this->floor(x), this->floor(y),
+                  this->floor(x + w) - int(scale()) - this->floor(x),
+                  this->floor(y + h) - int(scale()) - this->floor(y));
   }
 }
```

**Reply by dannye** [reply]
> About concentric rectangles: you could try this to draw a 2 pixel-thick rectangle:
> 
> ```
> fl_line_style(FL_SOLID,2); 
> fl_rect(x,y,w,h);
> fl_line_style(FL_SOLID);
> ```

FTR, I was able to fix all "concentric rectangle" issues well enough using this approach for 2-pixel and 3-pixel borders by adding offsets to x/y/w/h as needed to get the desired outcome. ([link](https://github.com/dannye/crystal-tracker/commit/0602c35cf1bc8cac3a2428ffb12819d097771738))

The only remaining issue is the blocky icons (assuming I don't discover any new issues when I test again on Mac and Linux).

**Reply by ManoloFLTK** [reply]
> The only remaining issue is the blocky icons 

I find this procedure to produce a decent pencil icon at 125% :
```
  Fl_Button b_pixmap(x,y,w,h,"Pixmap");
  Fl_Pixmap *pixmap = new Fl_Pixmap(PENCIL_XPM);
  int W = pixmap->w(), H = pixmap->h();
  Fl_RGB_Image *pixmap2 = new Fl_RGB_Image(pixmap);
  delete pixmap;
  Fl_Image::RGB_scaling(FL_RGB_SCALING_BILINEAR);
  Fl_RGB_Image *pixmap3 = (Fl_RGB_Image*)pixmap2->copy(2 * W, 2 * H);
  Fl_Image::RGB_scaling(FL_RGB_SCALING_NEAREST);
  delete pixmap2;
  pixmap3->scale(W, H);
  b_pixmap.image(pixmap3);
```

**Reply by dannye** [reply]
This is a good improvement, thank you. Is there a callback I can use for when FLTK's internal scaling factor is changed, so that I can use the original xpm when at 100% and use the RGB image otherwise?

**Reply by ManoloFLTK** [reply]
I don't think there's a callback _per se_. A workaround would be
- memorize in your widget's draw function the current value of the scaling factor as follows:
```
  int nscreen = widget->window()->screen_num();  // the screen where window containg widget is mapped
  float s = Fl::screen_scale(nscreen); // this screen's scale factor
```
- detect when this value changes from or to 1.

**Reply by dannye** [reply]
Thanks. This is looking really promising, but the final result is missing the right-most column and bottom-most row of pixels:
![image](https://github.com/user-attachments/assets/419ff008-0718-46fd-b5bc-ad54710d8030)

The original xpm looks like this:
![image](https://github.com/user-attachments/assets/1c79304f-66c3-416b-a8f8-a88669852c56)
https://github.com/dannye/crystal-tracker/blob/main/res/snip.xpm

Is this possibly a minor bug with [Fl_RGB_Image::copy()](https://github.com/fltk/fltk/blob/master/src/Fl_Image.cxx#L467)?

**Reply by dannye** [reply]
Sorry, I should have done a little more testing first. I do not think this is an issue with Fl_RGB_Image::copy(). I only see the cropping issue when the Windows scale is scaled up. (The copied/scaled RGB image looks good when the Windows scale is at 100%.)

**Reply by ManoloFLTK** [reply]
OK. I can see that. The cure I suggest is to make sure  the .xpm file contains at least one row/column of transparent pixels on each of its 4 sides. When that's not the case, as with the right and bottom sides of `snip.xpm`, manually add one such row/column in the .xpm file. I did that (attached) and the scissors icon display becomes correct after that.
[snip-17.xpm.txt](https://github.com/user-attachments/files/17389510/snip-17.xpm.txt)

Of course, the best approach to render an xpm-using app HighDPI-compatible is to somehow construct larger xpm's.

**Reply by dannye** [reply]
Okay, I have a few dozen to update but I'll just bite the bullet and add the padding to them. Thanks again for everything, your help means a lot. I think that solves everything! Sorry if my hyperfixation makes me annoying at times!

**Comment by ManoloFLTK** [comment]
Please, could you post here the source file of your pencil xpm ? I will experiment with how to best scale it.

**Reply by dannye** [reply]
Yes, definitely: https://github.com/dannye/crystal-tracker/blob/main/res/pencil.xpm

**Reply by ManoloFLTK** [reply]
Thanks for this xpm icon. I've looked into this problem and tried to first scale it up to twice its size by Fl_Pixmap::copy() before scaling it down with Fl_Pixmap::copy() again to the desired size. This brings no improvement.

I see now 2 possibilities for you.
1) the preferred one would be to construct a true, clean pencil.xpm twice larger (32x32) than the current pencil icon (16x16), give this new icon a drawing size of 16x16 with `big_pencil->scale(16,16);` and assign this new icon to your Fl_Button or Fl_Box just as before.

or

2) keep using your 16x16 pixel-icon but transform it (at run time) into an Fl_RGB_Image before assigning it as a Fl_Box or Fl_Button's image:
```
Fl_Pixmap *pencil = new Fl_Pixmap("PENCIL_XPM");
Fl_RGB_Image *rgb_pencil = new Fl_RGB_Image(pencil);
delete pencil;
Fl_Box *box = ...
box->image(rgb_pencil);
```

**Reply by dannye** [reply]
This was the result of trying out option 2:

![image](https://github.com/user-attachments/assets/10b91c45-a092-4ecf-bc09-7c489a432350)

**Comment by MatthiasWM** [comment]
Sorry, I can currently check only macOS rendering. I used Affinity Photo to upscale your pixel graphics. There is a special function for that, so that scaled down graphics looks like the original. Here are the scaled up versions:
![pencil](https://github.com/user-attachments/assets/d7470089-8e62-48c9-be7d-fd2b30378bce)
![pencil_x2](https://github.com/user-attachments/assets/833ae822-f78b-44ff-a6e8-dc16a9d0d046)
![pencil_x3](https://github.com/user-attachments/assets/07d2bbae-ba8a-47e1-921e-68969b8e8f5a)
![pencil_x4](https://github.com/user-attachments/assets/08c73078-7775-455b-bc13-6ad17403ec60)

Using pencil_x2 I get the following graphics for various scales between 50% and 110% (I have a HiDef display, so you 130% corresponds to the second image here, I guess):
![Screenshot 2024-10-14 at 19 17 05](https://github.com/user-attachments/assets/d4fb1554-88cc-4d31-8fc7-90ac9217a642)
![Screenshot 2024-10-14 at 19 17 41](https://github.com/user-attachments/assets/f20c3946-05a3-4176-be39-84fd6d4e5afd)
![Screenshot 2024-10-14 at 19 17 51](https://github.com/user-attachments/assets/36d4d35c-67be-45e5-a1d1-3dac3884aae4)
![Screenshot 2024-10-14 at 19 18 04](https://github.com/user-attachments/assets/7beaa5da-422c-450d-bca2-66263c3d009b)
![Screenshot 2024-10-14 at 19 18 27](https://github.com/user-attachments/assets/5baeb9e9-cbb8-49a8-a5a0-010950dd05c4)

This last one was done by scaling down pencil_x3:
![Screenshot 2024-10-14 at 19 19 17](https://github.com/user-attachments/assets/db2c92b5-e122-4c94-8aa0-7e93aa023143)

Sorry for Apple's antialiased snapshot function, but the last one at 33% percent looks pretty much exactly like the original 5pm.

```
        { Fl_Button* o = new Fl_Button(155, 281, 24, 24);
          o->image( image_pencil_x2() );
          o->image()->scale(16, 16, 0, 1);
          o->labelsize(11);
        } // Fl_Button* o
```

Edit: ok, I just got to test it on a Windows HiDef display and the results are not so nice but also not as bad as some of the other snapshots I have seen here. Maybe my setup uses a better `SetStretchBltMode` for `StretchBlt` ?

**Reply by dannye** [reply]
Scaling up in such a way that it looks like the original when scaled back down is definitely interesting. I will keep this in mind, thanks.

**Comment by ManoloFLTK** [comment]
This new patch (remove first all previous changes to the FLTK lib) should fix all your issues with filling boxes and keeping them adjacent : 

[rectboundV4.patch.txt](https://github.com/user-attachments/files/17376756/rectboundV4.patch.txt)

**Reply by dannye** [reply]
This looks great to me. All expected issues are resolved and I see no other side effects. Thank you very much for all your help.

**Reply by ManoloFLTK** [reply]
The equivalent of this patch, still simpler yet, is now committed to FLTK's git repo.

**Comment by ManoloFLTK** [comment]
Regarding your nested rectangles of diverse colors: an approach robust to scaling is possible if the content of these rectangles can be drawn after them. The trick is to draw nested filled rectangles. Replace, for example,
```
fl_color(FL_BLACK); fl_rect(x,y,w,h);
fl_color(FL_RED); fl_rect(x+1,y+1,w-2,h-2);
fl_color(FL_BLACK); fl_rect(x+2,y+2,w-4,h-4);
```
by
```
fl_color(FL_BLACK); fl_rectf(x,y,w,h);
fl_color(FL_RED); fl_rectf(x+1,y+1,w-2,h-2);
fl_color(FL_BLACK); fl_rectf(x+2,y+2,w-4,h-4);
fl_color(background); fl_rectf(x+3,y+3,w-6,h-6);
```
and draw what goes inside these rectangles after.

**Reply by dannye** [reply]
This is probably better than the current solution of setting the line style width to 2 or 3 pixels, because it probably eliminates the chances of drawing outside the box's expected area, at the cost of a little more overdraw. I will tinker with this, thank you.

**Comment by dannye** [comment]
I found a much more satisfying solution for the blocky/blurry icon images: Use `fl_override_scale()` in order to choose the largest integer scale that allows the image to still fit inside the button.

100%:

![image](https://github.com/user-attachments/assets/aec9df73-feae-4b8e-b1d5-b0b9c8b1cb34)

125%:

![image](https://github.com/user-attachments/assets/410b77e8-55a6-4574-9922-a6c4660b89fc)

150%:

![image](https://github.com/user-attachments/assets/af37774d-c3d0-4401-99f0-7feca9769b28)

175%:

![image](https://github.com/user-attachments/assets/5544cff1-5fc8-4ac8-ad20-0e2340c25b22)

This was done by adding the following to my custom button's `draw()` override:

```cpp
#ifndef _WIN32
	draw_label();
#else
	Fl_Image *img = active_r() ? image() : deimage();
	float scale = fl_override_scale();
	int SX = (int)(x() * scale + (x() < 0 ? -0.001f : 0.001f));
	int SY = (int)(y() * scale + (y() < 0 ? -0.001f : 0.001f));
	int SW = (int)(w() * scale + (w() < 0 ? -0.001f : 0.001f));
	int SH = (int)(h() * scale + (h() < 0 ? -0.001f : 0.001f));
	int IW = img->w();
	int IH = img->h();
	if (SW < IW || SH < IH) {
		fl_restore_scale(scale);
		draw_label();
		return;
	}
	int W = IW * (SW / IW);
	int H = IH * (SH / IH);
	int X = (int)(SX + SW/2.0f - W/2.0f + 0.5f);
	int Y = (int)(SY + SH/2.0f - H/2.0f + 0.5f);
	img->scale(W, H, 0, 1);
	img->draw(X, Y, W, H, 0, 0);
	img->scale(IW, IH, 0, 1);
	fl_restore_scale(scale);
#endif
```
